dartagnan/src/main/java/com/dat3m/dartagnan/expression/type/AggregateType.java

```diff
@@ -1,23 +1,29 @@
 package com.dat3m.dartagnan.expression.type;
 
-import com.dat3m.dartagnan.expression.Type;
-
 import java.util.List;
 import java.util.stream.Collectors;
 import java.util.stream.IntStream;
 
+import com.dat3m.dartagnan.expression.Type;
+
 public class AggregateType implements Type {
 
     private final List<TypeOffset> directFields;
+    private final List<Type> fields;
 
     AggregateType(List<Type> fields, List<Integer> offsets) {
         this.directFields = IntStream.range(0, fields.size()).boxed().map(i -> new TypeOffset(fields.get(i), offsets.get(i))).toList();
+        this.fields = fields;
     }
 
     public List<TypeOffset> getTypeOffsets() {
         return directFields;
     }
 
+    public List<Type> getFields() {
+        return fields;
+    }
```

Comment:
There is no need to store the types separately.
Just use `Lists.transform(directFields, x -> x.type())`.